astropy/convolution/boundary_none_direct.c

```diff
@@ -0,0 +1,449 @@
+        const size_t * const image_shape,
+        const DTYPE * const g,
+        const size_t * const kernel_shape,
+        const bool nan_interpolate,
+        const unsigned n_threads);
+
+// 1D
+void convolve1d_boundary_none_c(DTYPE * const result,
+        const DTYPE * const f, const size_t nx,
+        const DTYPE * const g, const size_t nkx,
+        const bool nan_interpolate,
+        const unsigned n_threads);
+inline __attribute__((always_inline)) void convolve1d_boundary_none(DTYPE * const result,
+        const DTYPE * const f, const size_t nx,
+        const DTYPE * const g, const size_t nkx,
+        const bool nan_interpolate,
+        const unsigned n_threads);
+// 2D
+void convolve2d_boundary_none_c(DTYPE * const result,
+        const DTYPE * const f, const size_t nx, const size_t ny,
+        const DTYPE * const g, const size_t nkx, const size_t nky,
+        const bool nan_interpolate,
+        const unsigned n_threads);
+inline __attribute__((always_inline)) void convolve2d_boundary_none(DTYPE * const result,
+        const DTYPE * const f, const size_t nx, const size_t ny,
+        const DTYPE * const g, const size_t nkx, const size_t nky,
+        const bool nan_interpolate,
+        const unsigned n_threads);
+// 3D
+void convolve3d_boundary_none_c(DTYPE * const result,
+        const DTYPE * const f, const size_t nx, const size_t ny, const size_t nz,
+        const DTYPE * const g, const size_t nkx, const size_t nky, const size_t nkz,
+        const bool nan_interpolate,
+        const unsigned n_threads);
+inline __attribute__((always_inline)) void convolve3d_boundary_none(DTYPE * const result,
+        const DTYPE * const f, const size_t nx, const size_t ny, const size_t nz,
+        const DTYPE * const g, const size_t nkx, const size_t nky, const size_t nkz,
+        const bool nan_interpolate,
+        const unsigned n_threads);
+
+void convolveNd_boundary_none_c(DTYPE * const result,
+        const DTYPE * const f,
+        const unsigned n_dim,
+        const size_t * const image_shape,
+        const DTYPE * const g,
+        const size_t * const kernel_shape,
+        const bool nan_interpolate,
+        const unsigned n_threads)
+{
+    if (!result || !f || !g || !image_shape || !kernel_shape)
+            return;
+
+    if (n_dim == 1)
+        convolve1d_boundary_none_c(result, f,
+                image_shape[0],
+                g, kernel_shape[0],
+                nan_interpolate, n_threads);
+    else if (n_dim == 2)
+        convolve2d_boundary_none_c(result, f,
+                image_shape[0], image_shape[1],
+                g, kernel_shape[0], kernel_shape[1],
+                nan_interpolate, n_threads);
+    else if (n_dim == 3)
+        convolve3d_boundary_none_c(result, f,
+                        image_shape[0], image_shape[1], image_shape[2],
+                        g, kernel_shape[0], kernel_shape[1], kernel_shape[2],
+                        nan_interpolate, n_threads);
```

Comment:
> Raise an error in in the else case

Is this absolutely necessary? The C functions aren't intended to be called directly and `convolve` checks the dims and raises as desired.

I ask now, as this may not be elegant to implement (if at all possible). Trivially, it can `assert(0)` such that there's something in place, however, this will still be silent when triggered within a notebook. It would also have to undefine `NDEBUG`, not that this is an issue. BTW I'm looking into why `-NDEBUG` is passed as a compiler flag, were/are you aware that it is, as there are 113 `asserts()` used in astropy C code?